membership/src/main/scala/zio/membership/hyparview/plumtree/PlumTree.scala

```diff
@@ -0,0 +1,188 @@
+      onReceive: Either[UUID, (UUID, T)] => UIO[Unit]
+    ): ZIO[PeerState[T] with Logging with PeerService[T], membership.Error, Unit] =
+      PeerService.receive[T].foreach {
+        case (sender, msg) =>
+          msg match {
+            case Prune =>
+              PeerState.moveToLazyPeers(sender).commit
+            case Graft(uuid) =>
+              deduplication.get(uuid).commit.flatMap {
+                case None => logWarn(s"Expected to find message $uuid but didn't")
+                case Some(msg) =>
+                  PeerService
+                    .send(sender, msg)
+                    .foldCauseM(
+//                      e => log.error(s"Failed replying to Graft message from $sender", e),
+                      logError,
+                      _ => logDebug(s"Replied to Graft message from $sender")
+                    ) *> PeerState.addToEagerPeers(sender).commit
+              }
+            case IHave(uuids) =>
+              def process(uuid: UUID) =
+                deduplication.contains(uuid).commit.flatMap {
+                  case true =>
+                    logDebug(s"Received IHave for already received message $uuid")
+                  case false =>
+                    onReceive(Right((uuid, sender)))
+                }
+              ZIO.foreach_(uuids)(process)
+            case UserMessage(payload) =>
+              userOut(payload) <* logDebug(s"Received userMessage from $sender")
+            case m @ Gossip(uuid, payload) =>
+              deduplication.add(uuid, m).commit.flatMap {
+                case true =>
+                  // first occurrence
+                  onReceive(Left(uuid)) *>
+                    (PeerState
+                      .addToEagerPeers(sender) *> PeerState.eagerPushPeers[T].zip(PeerState.lazyPushPeers[T])).commit
+                      .flatMap {
+                        case (eagerPeers, lazyPeers) =>
+                          val sendEager = ZIO.foreach_(eagerPeers.filterNot(_ == sender)) { peer =>
+                            PeerService
+                              .send(peer, m)
+                              .foldCauseM(
+//                                e => log.error(s"Failed sending message to $peer", e),
+                                logError,
+                                _ => logInfo(s"Sent Gossip to $peer")
+                              )
+                          }
+
+                          // no need to filter here as we ensured sender is in eager peers
+                          val sendLazy = ZIO.foreach_(lazyPeers.map((_, uuid)))(gossipOut.tupled)
+
+                          sendEager *> sendLazy *> userOut(payload)
+                      }
+                case false =>
+                  // seen before -> disconnect sender
+                  PeerService
+                    .send(sender, Prune)
+                    .foldCauseM(
+//                      e => logError(s"Failed to send prune message to $sender", e),
```

Comment:
when we merge this to master we need to migrate `zio-logging` I added `log.error(msg, error)`

Reply:
many thanks 🙇

Reply:
99 compile errors. 😱

Reply:
done